### DISCUSSION TITLE
Which version of OpenSSL might the ENGINE_load_private_key function be deprecated? Or in which year might it be deprecated?

### DISCUSSION THREAD

**Discussion by davecheng2024** [openssl/openssl#23310] [Q&A] [answered]
Engine-backed keys can be loaded via custom OSSL_STORE implementation. In this case the EVP_PKEY objects created by ENGINE_load_private_key() is expected to be deprecated in which version of OpenSSL? Or which year version will it be deprecated？Or OpenSSL will always support this function？

**Comment by t8m** [comment, marked as answer]
Engine backed keys were deprecated in the 3.0 release. The support will not be removed earlier than in the 4.0 release. However it is not yet decided whether it will be removed there or even later. There is also no date set for the 4.0 release - however it can fairly safely be said that it will not be released earlier than in 2025.

**Reply by davecheng2024** [reply]
Thank you for your response. I will ask our developers to replace the engine with the provider in 2025.

**Comment by petrovr** [comment]
davecheng2024 wrote:
> Engine-backed keys can be loaded via custom OSSL_STORE implementation. In this case the EVP_PKEY objects created by ENGINE_load_private_key() is expected to be deprecated in which version of OpenSSL? Or which year version will it be deprecated?
>

Not exactly as OpenSSL Store2 API deprecates "engine" related functions in 3+.

ENGINE_load_...(1) and STORE...load...(2) could share code but second one is deprecated as well.

Dunno about provider related Store 2 API or maybe is better to call it Store 3 API.

Regards,
Roumen Petrov

**Reply by davecheng2024** [reply]
Thank you for your response. I will ask our developers to replace the engine with the provider in 2025.
